Fix synchronize retries

cboylan opened the pull request:
#### ISSUE TYPE

- bugfile pullrequest

#### ANSIBLE VERSION

2.3

#### SUMMARY

The synchronize module munges its task args on every invocation of
run(). This was problematic because the munged data was not fit for use
by a second pass of the synchronize module. Correct this by using a copy
of the task args on every invocation of run() so that the original args
are not affected.

Local testing using this playbook seems to confirm that things work as
expected:

```
  - hosts: all
  tasks:
    - delay: 2
      register: task_result
      retries: 1
      until: task_result.rc == 0
      synchronize:
        dest: /tmp/out
        mode: pull
        src: /tmp/nonexistent/
```

fixes #18281

cboylan committed:
Fix synchronize retries

The synchronize module munges its task args on every invocation of
run(). This was problematic because the munged data was not fit for use
by a second pass of the synchronize module. Correct this by using a copy
of the task args on every invocation of run() so that the original args
are not affected.

Local testing using this playbook seems to confirm that things work as
expected:

  - hosts: all
  tasks:
    - delay: 2
      register: task_result
      retries: 1
      until: task_result.rc == 0
      synchronize:
        dest: /tmp/out
        mode: pull
        src: /tmp/nonexistent/

fixes #18281

cboylan committed:
Update synchroncization fixture assertions

When we started operating on a copy of the task args the test assertions
were no longer asserting things about the munged state but of the
pristine state. Convert the copy of task args to a class member so that
it can be compared against later in testing and update the assertions to
check this munged copy.

cboylan committed:
Fix missing . in object path

Its self._tmp_task_args not self_tmp_task_args.

cboylan committed:
Shuffle objects around for cleaner testing

Attach the temporary args dict to the task rather thant the action as
this makes updating the existing tests cleaner.

jctanner commented:
@tima can you take a look?

tima commented:
@jctanner Will put it on my TODO list. Looks good but I haven't tested it. We should also have @abadger look at this.

abadger commented on lib/ansible/plugins/action/synchronize.py:
```diff
@@ -155,6 +155,11 @@ def run(self, tmp=None, task_vars=None):
+        # We make a copy of the args here because we may fail and be asked to
+        # retry. If that happens we don't want to pass the munged args through
+        # to our next invocation. Munged args are single use only.
+        self._task._tmp_args = self._task.args.copy()
```
_tmp_args should be a local variable rather than an instance attribute.  But if we do that the unittests will need changing in a more intrusive way.  I think that's the right choice but I haven't looked at how to achieve that yet.

bcoca replied:
we should already be inside a copy of the task, not sure why we would need to copy args again.

cboylan replied:
The problem as I understand it is this _task object is reused on every retry. So while it may be a copy the first time around it isn't a fresh copy each time.

bcoca replied:
if that is an issue for the test, the test might not be correct, if this is an issue in general, it should be fixed in the code.

cboylan replied:
Its not an issue with the test, its an issue with the code.... All of this documented in the issue this fixes. https://github.com/ansible/ansible/issues/18281#issuecomment-261382356 specifically addresses this. I agree it should be fixed more generally, but that seems like a relatively massive change to ansible itself, and it was not the fix suggested to me. I doubt I would be able to work through the larger fix.

abadger replied:
Yeah, this is a quick fix that I understand will work.  I'm not certain whether changing retries to create a new action plugin on every iteration will have unwanted (or just buggy) side-effects.  I'd merge this and cherry-pick to 2.2.  Then if we do decide that retries can be changed, we can make that change in devel only.

jimi-c replied:
I agree, this should not be an attribute.

abadger commented on test/units/plugins/action/test_synchronize.py:
```diff
@@ -99,11 +99,11 @@ class SynchronizeTester(object):
 
 
-    def _execute_module(self, module_name, task_vars=None):
+    def _execute_module(self, module_name, module_args=None, task_vars=None):
```
Is this change needed?  It doesn't seem wrong, necessarily... I'm just wondering whether it is necessary and if so if there's something that should be refactored at a later date.

abadger replied:
Oh I see... this is a fake _execute_module() in the test case.  It's capturing information about how it was called.

So I think the way to adjust the test cases is to set ```self.final_module_args = module_args``` in this function and then adjust all of the assertions to use self.final_module_args instead of self.task._tmp_args.

bcoca replied:
not sure if the test should modify the function signature ... we can end up with test code that would never work with the real thing ... which makes me question the tests.

cboylan replied:
_execute_module has a signature of:

```lib/ansible/plugins/action/__init__.py:    def _execute_module(self, module_name=None, module_args=None, tmp=None, task_vars=None, persist_files=False, delete_remote_tmp=True):```

The tests already use an incomplete function signature that is being updated to accept the explicit module_args rather than the implicit args via the task object.

I can update to make the tmp args function local and check against the args passed into this function, but any issue with the tests was preexisting.

abadger replied:
Agreed with @cboylan here.  This is the pre-existing way that the test code works.  If we want to change it to have a more complete signature we should probably do that as a separate change.

abadger commented:
Made some suggestions.  I think it looks good otherwise.  @cboylan I also read your comments in the bug report.  It sounds good to me but I'm also not familiar enough with that code to know if something depends on reusing the AnsibleModule.  I'll ping jimi and bcoca about whether the design could be changed the way you propose.

We're going to be in face-to-face meetings this upcoming week so lots of travelling.  If you update and I don't respond feel free to ping me on irc (I'm abadger1999 in #ansible-devel on irc.freenode.net) to get it merged.

cboylan commented:
Ok so it sounds like I should go ahead and make the suggested updates and we can go from there. I too have had a busy week so only just swinging back around to this. Hopefully have code pushed later today.

cboylan committed:
Use function local args

This updates the fix to use a function local tmp args set that can be
munged. Tests are updated to remember what these values are when passed
in so that they can be compared against.

cboylan commented:
The shippable test failure appears to be because lib/ansible/modules/core and lib/ansiblemodules/extras are missing __init__.py files. I didn't touch either path so not sure that those are related to this pull request.

tima commented:
I think it may have something to do with the recent repo  merge @cboylan. Sorry about that. I will try and remember to ping @abadger @jctanner and @jimi-c after the weekend if they don't see this themselves.

abadger commented:
Thanks @cboylan !  This has been merged to devel for inclusion in the 2.3 release.

chouseknecht commented:
@abadger merging this PR busted synchronize over docker connection plugin. Fixed the issue with #19531
